tls/s2n_ktls_io.c

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <sys/socket.h>
+
+#include "tls/s2n_ktls.h"
+#include "utils/s2n_socket.h"
+
+#if S2N_KTLS_SUPPORTED /* CMSG_* macros are platform specific */
+
+/*
+ * sendmsg and recvmsg are syscalls which can be used to send 'real' data along
+ * with ancillary data. Ancillary data is used to communicate to the socket the
+ * type of the TLS record being sent/received.
+ *
+ * Ancillary data macros (CMSG_*) are platform specific and gated.
+ */
+
+S2N_RESULT s2n_ktls_send_msg_impl(int sock, struct msghdr *msg,
+        struct iovec *msg_iov, size_t count, s2n_blocked_status *blocked, ssize_t *send_len)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(msg_iov);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(send_len);
+    RESULT_ENSURE_GT(count, 0);
+
+    /* set send buffer */
+    msg->msg_iov = msg_iov;
+    msg->msg_iovlen = count;
+
+    *blocked = S2N_BLOCKED_ON_WRITE;
+    *send_len = sendmsg(sock, msg, 0);
+    if (*send_len < 0) {
+        if (errno == EWOULDBLOCK || errno == EAGAIN) {
+            RESULT_BAIL(S2N_ERR_IO_BLOCKED);
+        }
+        RESULT_BAIL(S2N_ERR_IO);
+    }
+    *blocked = S2N_NOT_BLOCKED;
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_set_ancillary_data(struct msghdr *msg, uint8_t record_type)
+{
+    RESULT_ENSURE_REF(msg);
+
+    RESULT_ENSURE_REF(msg->msg_control);
+    RESULT_ENSURE_GTE(msg->msg_controllen, CMSG_SPACE(sizeof(uint8_t)));
+
```

Comment:
Maybe nit, but using sizeof(uint8_t) everywhere might not exactly capture your requirements. Maybe replace every call to sizeof(uint8_t) with sizeof(record_type)?